File: apps/tauri/src-tauri/src/types/export.rs
//! Export types
//!
//! Types for data and schema export operations.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ExportFormat {
    Csv,
    Json,
    Sql,
    Xlsx,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportDataRequest {
    pub connection_id: String,
    pub table: String,
    pub format: ExportFormat,
    pub file_path: String,
    pub schema: Option<String>,
    pub columns: Option<Vec<String>>,
    pub include_headers: Option<bool>,
    pub delimiter: Option<String>,
    pub pretty_print: Option<bool>,
    pub sheet_name: Option<String>,
    pub rows: Option<Vec<HashMap<String, serde_json::Value>>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportDataResponse {
    pub success: bool,
    pub rows_exported: Option<u64>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportSchemaRequest {
    pub connection_id: Option<String>,
    pub snapshot_id: Option<String>,
    pub file_path: Option<String>,
    pub format: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportSchemaResponse {
    pub success: bool,
    pub file_path: Option<String>,
    pub error: Option<String>,
}

File: apps/electron/src/main/ipc/handlers/export.ts
/**
 * Export IPC Handler
 *
 * Handles all export-related IPC operations.
 */

import type {
  ColumnInfo,
  ExportBundleRequest,
  ExportComparisonReportRequest,
  ExportRequest,
  ExportSchemaRequest,
  SchemaComparisonResult,
} from '@shared/types';
import type { Buffer } from 'node:buffer';
import type {HandlerContext} from '../base/handler';
import {
  generateHTMLReport,
  generateJSONReport,
  generateMarkdownReport,
} from '@/lib/comparison-report-generators';
import {
  generateCSV,
  generateExcel,
  generateJSON,
  generateSQL,
} from '@/lib/export-generators';
import { databaseService } from '../../services/database';
import {
  exportBundle,
  exportSchema,
} from '../../services/query-schema-sharing';
import { getSchemaSnapshot } from '../../services/store';
import {  IpcHandler } from '../base/handler';

export class ExportHandler extends IpcHandler {
  constructor() {
    super({ name: 'export' });
  }

  register(): void {
    this.handleLegacy('export:data', this.exportData.bind(this));
    this.handleLegacy('export:bundle', this.exportBundle.bind(this));
    this.handleLegacy('export:schema', this.exportSchema.bind(this));
    this.handleLegacy(
      'export:comparison-report',
      this.exportComparisonReport.bind(this)
    );
  }

  private async exportData(
    request: ExportRequest,
    _ctx: HandlerContext
  ): Promise<{ data: Buffer | string; format: string }> {
    let columns: ColumnInfo[] = [];
    let rows: Record<string, unknown>[] = [];

    const tableData = await databaseService.getTableData(
      request.connectionId,
      request.table,
      1,
      999999,
      undefined,
      undefined,
      undefined,
      request.schema
    );

    if (tableData.success) {
      columns = tableData.columns || [];
      rows = (tableData.rows || []) as Record<string, unknown>[];
    }

    let output: Buffer | string = '';
    if (request.format === 'csv') {
      output = generateCSV(rows, columns);
    } else if (request.format === 'json') {
      output = generateJSON(rows, columns);
    } else if (request.format === 'sql') {
      output = generateSQL(rows, columns, { tableName: request.table });
    } else if (request.format === 'xlsx') {
      output = await generateExcel(rows, columns, {
        sheetName: request.table,
      });
    }

    return {
      data: output,
      format: request.format,
    };
  }

  private async exportBundle(
    _request: ExportBundleRequest,
    _ctx: HandlerContext
  ): Promise<{ success: boolean; data: unknown }> {
    const { data } = await exportBundle({
      name: 'bundle',
      queries: [],
      schemas: [],
    });
    return { success: true, data };
  }

  private async exportSchema(
    request: ExportSchemaRequest,
    _ctx: HandlerContext
  ): Promise<{ success: boolean; data: unknown }> {
    if (request.snapshotId) {
      const snapshot = getSchemaSnapshot(request.snapshotId);
      if (!snapshot) {
        throw new Error(
          `Schema snapshot with ID ${request.snapshotId} not found`
        );
      }
      const { data } = await exportSchema({
        name: snapshot.name,
        description: snapshot.description,
        databaseName: snapshot.connectionPath || '',
        databaseType: 'sqlite',
        format: 'json',
        schemas: snapshot.schemas || [],
      });
      return { success: true, data };
    }
    throw new Error('snapshotId is required');
  }

  private async exportComparisonReport(
    request: ExportComparisonReportRequest,
    _ctx: HandlerContext
  ): Promise<{ success: boolean; data: string; format: string }> {
    const comparisonData = (request.comparison ||
      request.comparisonResult) as SchemaComparisonResult;
    if (!comparisonData) {
      throw new Error('Comparison data is required');
    }
    let output = '';

    if (request.format === 'markdown') {
      output = generateMarkdownReport(comparisonData);
    } else if (request.format === 'html') {
      output = generateHTMLReport(comparisonData);
    } else if (request.format === 'json') {
      output = generateJSONReport(comparisonData);
    }

    return { success: true, data: output, format: request.format };
  }
}

// Export singleton instance
export const exportHandler = new ExportHandler();
